tool/src/org/antlr/v4/codegen/target/TypeScriptTarget.java

```diff
@@ -0,0 +1,204 @@
+			"module",
+			"new",
+			"null",
+			"number",
+			"package",
+			"private",
+			"protected",
+			"public",
+			"require",
+			"return",
+			"set",
+			"static",
+			"string",
+			"super",
+			"switch",
+			"symbol",
+			"this",
+			"throw",
+			"true",
+			"try",
+			"type",
+			"typeof",
+			"var",
+			"void",
+			"while",
+			"with",
+			"yield",
+			"of"
+	));
+
+	public TypeScriptTarget(CodeGenerator gen) {
+		super(gen);
+	}
+
+	@Override
+	protected Set<String> getReservedWords() {
+		return reservedWords;
+	}
+
+
+	@Override
+	public String getTargetStringLiteralFromANTLRStringLiteral(
+		CodeGenerator generator,
+		String literal, boolean addQuotes)
+	{
+		StringBuilder sb = new StringBuilder();
+		String is = literal;
+
+		if ( addQuotes ) sb.append('"');
+
+		for (int i = 1; i < is.length() - 1; ) {
+			int codePoint = is.codePointAt(i);
+			int toAdvance = Character.charCount(codePoint);
+			if  (codePoint == '\\') {
+				// Anything escaped is what it is! We assume that
+				// people know how to escape characters correctly. However
+				// we catch anything that does not need an escape in Java (which
+				// is what the default implementation is dealing with and remove
+				// the escape. The C target does this for instance.
+				//
+				int escapedCodePoint = is.codePointAt(i + toAdvance);
+				toAdvance++;
+				switch (escapedCodePoint) {
+					// Pass through any escapes that Java also needs
+					//
+					case    'n':
+					case    'r':
+					case    't':
+					case    'b':
+					case    'f':
+					case    '\\':
+						// Pass the escape through
+						sb.append('\\');
+						sb.appendCodePoint(escapedCodePoint);
+						break;
+
+					case    'u':    // Either unnnn or u{nnnnnn}
+						if (is.charAt(i + toAdvance) == '{') {
+							while (is.charAt(i + toAdvance) != '}') {
+								toAdvance++;
+							}
+
+							toAdvance++;
+						} else {
+							toAdvance += 4;
+						}
+
+						if (i + toAdvance <= is.length()) {
+							// we might have an invalid \\uAB or something
+							String fullEscape = is.substring(i, i + toAdvance);
+							appendUnicodeEscapedCodePoint(
+								CharSupport.getCharValueFromCharInGrammarLiteral(fullEscape),
+								sb);
+						}
+
+						break;
+
+					default:
+						if (shouldUseUnicodeEscapeForCodePointInDoubleQuotedString(escapedCodePoint)) {
+							appendUnicodeEscapedCodePoint(escapedCodePoint, sb);
+						} else {
+							sb.appendCodePoint(escapedCodePoint);
+						}
+
+						break;
+				}
+
+				// Go past the \ character
+				i++;
+			} else {
+				if (codePoint == 0x22) {
+					// ANTLR doesn't escape " in literal strings, but every other language needs to do so.
+					sb.append("\\\"");
+				} else if (shouldUseUnicodeEscapeForCodePointInDoubleQuotedString(codePoint)) {
+					appendUnicodeEscapedCodePoint(codePoint, sb);
+				} else {
+					sb.appendCodePoint(codePoint);
+				}
+			}
+
+			i += toAdvance;
+		}
+
+		if ( addQuotes ) sb.append('"');
+
+		return sb.toString();
+	}
```

Comment:
This method has a lot of in common with [`getTargetStringLiteralFromANTLRStringLiteral`](https://github.com/antlr/antlr4/blob/7e8c4f4d9b5659673f00289c1ee60c7a74007042/tool/src/org/antlr/v4/codegen/Target.java#L237) from the base class. Please merge logic.

Reply:
The issue actually is not resolved, I don't understand why it was marked as resolved.

Reply:
Because it's not an issue. Java and Javascript support slightly different literals.
See https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.10.5 and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String.
IMHO it's a bug to have just one translation method for all targets because inevitably users will complain, and we need to safely fix the translation on a per target basis.

Reply:
> IMHO it's a bug to have just one translation method for all targets because inevitably users will complain, and we need to safely fix the translation on a per target basis.

Firstly, it's not bug, but maybe incorrect implementation. Secondly, IMHO it's not optimal to create duplicated functionality because of slightly different things. They can be encapsulated to separated methods according to OOP. The less code we have, the less potential errors we have. Moreover, this method not only escapes chars but also joins them. Joining logic should be the same. For that matter, `charCount`, `codePointAt` and other char methods are Java-specific, they should not be used for TypeScript or other target literals, but actually it doesn't matter.

Reply:
Hi guys. @KvanTTT and I discussed having one method a while ago; is there a way to share common functionality but still have targets do special things when necessary?

Reply:
Wouldn't that create unnecessary complexity ?

Reply:
Actually, since the default inherited method sets escapeSpecial to false, the 2 methods end up behaving the same (they shouldn't since the list of escapes should be different)

Reply:
hmm...yeah, I guess I'm trying to avoid duplicating code, particularly since most targets will be similar. If they are sufficiently different then it warrants having separate sections of functionality. I think those sections used to be separate overridden functions but are now combined into a single function.  @KvanTTT can you remind me about the history of the combined function here? thanks!

Reply:
Sorry for the delay, I was figuring out of API of Target and how `getTargetStringLiteralFromString` and `getTargetStringLiteralFromANTLRStringLiteral` methods work. These methods are responsible for the conversion from grammar code (string literals or predicated) to target string literals. It looks like the single difference is the second method should conserve escaping characters since they are used in token names (that why `\n` from ANTLR string becomes `\\n` in target string).

I've clarified API of Target, fixed potential conversion bugs and removed duplication code in https://github.com/antlr/antlr4/pull/3891, please take a look.